specs/AddHostObjectToScriptWithOrigins.md

```diff
@@ -0,0 +1,275 @@
+# Background
+Currently WebView2 supports adding host objects to script from client application
+and then later those objects can be accessed from the JavaScript of the top level frame
+using `window.chrome.webview.hostObjects.{name}`.
+Documentation can be found [here](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2?view=webview2-1.0.705.50#addhostobjecttoscript).
+
+We were asked to support this feature in iframes including
+child frames that don't match the origin of the parent. In this document we describe API
+additions to support host objects in iframes. We'd appreciate your feedback.
+
+
+# Description
+The FrameCreated event is raised when an iframe is created and before starting any
+navigation and content loading. The FrameDeleted event is raised when an iframe is deleted
+or the document containing that iframe is destroyed.
+By providing event handler for frame created event,
+user can obtain the iframe and add host objects to it specifying the list of origins
+using `AddHostObjectToScriptWithOrigins` method.
+When script in the iframe tries to access a host object the origin of the iframe will be
+checked against the provided list. The Frame object has a Name property with the value of the
+name attribute from the iframe html tag declaring it. The NameChanged event is raised when
+corresponding iframe changes its window.name property.
+
+The host object will be removed during iframe deletion automatically.
+
+
+# Examples
+## Frame created event, frame name changed event, adding and removing host object
+
+## .NET, WinRT
+```c#
+void SubscribeToFrameCreated()
+{
+    webView.CoreWebView2.FrameCreated += delegate (object sender, CoreWebView2FrameCreatedEventArgs args)
+    {
+        if (args.Frame.Name.Equals("iframe_name"))
+        {
+            String[] origins = new String[] { "https://appassets.example" };
+            args.Frame.AddHostObjectToScriptWithOrigins("bridge", new BridgeAddRemoteObject(), origins);
+        }
+        args.Frame.NameChanged += delegate (object nameChangedSender, object nameChangedArgs)
+        {
+            CoreWebView2Frame frame = (CoreWebView2Frame)nameChangedSender;
+            if (!frame.Name.Equals("iframe_name"))
+            {
+                frame.RemoveHostObjectFromScript("bridge");
+            }
+        };
+    };
+}
```

Comment:
How synchronous is the NameChanged event? If I want to inject an object when the name matches the special name, is injecting in the NameChanged handler sufficient to deal with this:

```js
// Set my name to the magic value that gives me access to the bridge
window.name = "iframe_name";
// Go get it!
var bridge = chrome.webview.hostObjects.bridge; // does this succeed?
var token = await bridge.GetToken(); // ask the bridge for a token
```

or is there a race condition where the script needs to wait for the object to be injected

```js
// Set my name to the magic value that gives me access to the bridge
window.name = "iframe_name";
// Poll waiting for the bridge to show up
var check = async () => {
    var bridge = chrome.webview.hostObjects.bridge;
    if (!bridge) { setTimeout(check, 100); return; } // check again in 100ms
    var token = await bridge.GetToken(); // finally we can ask the bridge for a token
};
check();
```

Reply:
Additional point against adding host object to script based on name because name is async as mentioned.